torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,188 @@
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries.add(
+                TritonBundleEntry(
+                    kernel_hash, compile_time_ns, device, triton_cache_dir
+                )
+            )
+
+    @classmethod
+    def collect(
+        cls,
+    ) -> Tuple[List[TritonKernelArtifacts], Optional[TritonBundlerMetadata]]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return [], None
+
+        with dynamo_timed(
+            key="TritonBundler::collect", phase_name="TritonBundler", fwd_only=False
+        ):
+            if (entries := cls._entries) is not None:
+                result: List[TritonKernelArtifacts] = []
+                total_compile_time_ns = 0
+                for entry in entries:
+                    artifacts: List[TritonKernelArtifact] = []
+                    path = os.path.join(entry.directory, entry.kernel_hash)
+                    if not os.path.exists(path):
+                        continue
+                    for filename in os.listdir(path):
+                        filepath = os.path.join(path, filename)
+                        try:
+                            assert os.path.isfile(filepath)
+                            with open(filepath, "rb") as file:
+                                artifacts.append(
+                                    TritonKernelArtifact(filename, file.read())
+                                )
+                            counters["inductor"]["triton_bundler_save_kernel"] += 1
+                        except Exception:
+                            log.debug("failed to collect triton kernel", exc_info=True)
+                    if artifacts:
+                        result.append(
+                            TritonKernelArtifacts(
+                                entry.kernel_hash,
+                                entry.compile_time_ns,
+                                entry.device,
+                                artifacts,
+                            )
+                        )
+                        total_compile_time_ns += entry.compile_time_ns
+                cls._entries = None
+                num_kernels = len(result)
+                return result, TritonBundlerMetadata(num_kernels, total_compile_time_ns)
+            return [], None
+
+    @staticmethod
+    def write(bundle: List[TritonKernelArtifacts]) -> Optional[TritonBundlerMetadata]:
+        if not TritonBundler.is_enabled():
+            return None
+
+        with dynamo_timed(
+            key="TritonBundler::write", phase_name="TritonBundler", fwd_only=False
+        ):
+            num_kernels = 0
+            total_compile_time_ns = 0
+            basedir = os.getenv("TRITON_CACHE_DIR")
+            custom_dir = True
+            if basedir is None:
+                custom_dir = False
+                basedir = os.path.join(cache_dir(), "triton")
```

Comment:
This feels potentially duped with preexisting code somewhere? Is it possible to DRY this?

Reply:
There is a potentially silent performance problem that can happen when there bugs in this code, which goes like this: we accidentally put the Triton cache products in not quite the right place. Because there's no way to say "Triton, you MUST cache hit, error otherwise", Triton doesn't see the cache products and then goes and ahead and recompiles. This is completely silent and you only notice it because you are cache hit but it's still taking a long time to do Triton compilation.

The most obvious way to fix this is to somehow run Triton in a mode where cache hit is mandatory and error if it doesn't. But I don't know if we actually have an interposition point to do this (might need Triton change.) If we can't make it an error, then the other thing we can do is make sure we are carefully measuring compile time when we actually run Triton, and we can look for outliers WHEN bundled cache hit (it should be some low number and any time it is high there is a problem.) Unclear what logging you can have to further diagnose a problem like this though!

Reply:
> Is it possible to DRY this?

What does this mean?

Regarding changing Triton, it is something we can try negotiating with OAI, but likely will take a while

Reply:
DRY = Don't Repeat Yourself